Allow setup-env to specify Python version (#166)
cisagov/skeleton-generic: CONTRIBUTING.md

create and manage a Python virtual environment specific to this
project.

The `setup-env` tool in this repository is our recommended method
for automating the entire environment configuration process. The
dependencies required to run this tool are
[`gnu-getopt`](https://manned.org/getopt.1), `pyenv`, and
`pyenv-virtualenv`. If these tools are already configured on

> your system, you can simply run the following command.

Reviewer: I would like to clearly differentiate between GNU `getopt` the tool and `gnu-getopt` the Homebrew formula. Also updating the URL so that `getopt` points at the source repository and `pyenv` and `pyenv-virtualenv` point at their respective repositories. Updated the ending punctuation per suggestion of @jsf9k.

```suggestion
[GNU `getopt`](https://github.com/util-linux/util-linux/blob/master/misc-utils/getopt.1.adoc),
[`pyenv`](https://github.com/pyenv/pyenv), and [`pyenv-virtualenv`](https://github.com/pyenv/pyenv-virtualenv).
If these tools are already configured on your system, you can simply run the
following command:
```

Reviewer 2: @mcdonnnj - I think the correct punctuation after "...you can simply run the following command" is a colon, not a period.

Reviewer: @jsf9k Ah, I can add that to the suggestion. I wasn't paying attention to what came after this blurb and left the ending punctuation as-is.

create and manage a Python virtual environment specific to this
project.

> We recommend using the `setup-env` tool located in this repository,

Reviewer: Just a preference.

```suggestion
We recommend using the `setup-env` script located in this repository,
```

> for automating the entire environment configuration process. The

Reviewer 2: Slightly less awkward:
```suggestion
We recommend using the `setup-env` tool located in this repository,
as it automates the entire environment configuration process. The
```

> dependencies required to run this tool are

Reviewer: In line with the previous language change suggestion.

```suggestion
dependencies required to run this script are
```

Otherwise, follow the steps below to manually configure your
environment.

#### Installing and using `gnu-getopt`, `pyenv`, and `pyenv-virtualenv` ####

On the Mac, we recommend installing [brew](https://brew.sh/).  Then
installation is as simple as `brew install gnu-getopt pyenv pyenv-virtualenv` and
adding this to your profile:

```bash

> export PATH="/usr/local/opt/gnu-getopt/bin:$PATH"
> export PYENV_ROOT="$HOME/.pyenv"
> export PATH="$PYENV_ROOT/bin:$PATH"

Reviewer 2: No need for to separate the `PATH` additions:
```suggestion
export PYENV_ROOT="$HOME/.pyenv"
export PATH="$PYENV_ROOT/bin:/usr/local/opt/gnu-getopt/bin:$PATH"
```

Otherwise, follow the steps below to manually configure your
environment.

> #### Installing and using `gnu-getopt`, `pyenv`, and `pyenv-virtualenv` ####

Reviewer: Again clearly separating the actual requirement from the Homebrew formula.

```suggestion
#### Installing and using GNU `getopt`, `pyenv`, and `pyenv-virtualenv` ####
```

On the Mac, we recommend installing [brew](https://brew.sh/).  Then
installation is as simple as `brew install gnu-getopt pyenv pyenv-virtualenv` and
adding this to your profile:

```bash
export PYENV_ROOT="$HOME/.pyenv"

> export PATH="$PYENV_ROOT/bin:/usr/local/opt/gnu-getopt/bin:$PATH"

Reviewer 3: Brew on my system defaults to installing in `/opt/homebrew`, not `/usr/local`.  You should use `brew --prefix` instead:
```suggestion
export PATH="$PYENV_ROOT/bin:$(brew --prefix)/opt/gnu-getopt/bin:$PATH"
```

> export PATH="$PYENV_ROOT/bin:$(brew --prefix)/opt/gnu-getopt/bin:$PATH"

Reviewer: This should already be handled by correctly installing Homebrew. Any Homebrew formulae I have installed work without this kind of additional PATH modification.

```suggestion
export PATH="$PYENV_ROOT/bin:$PATH"
```

Reviewer 3: On my system, if I don't add that blurb (`$(brew --prefix)/opt/gnu-getopt/bin`) to my `PATH`, it uses the version of `getopt` provided by macOS:

```
# Default state
❱ which getopt
/usr/bin/getopt

# After PATH update
❱ which getopt
/opt/homebrew/opt/gnu-getopt/bin/getopt
```

Reviewer: Agh you are right. It's keg-only on macOS and I have a file in my `.bashrc.d/` that adds it to the path with other keg-only formulae that I want to use. Fixing the PATH for `getopt` should be separate from `pyenv` environment configuration so this suggestion is still something I want, but it should be moved instead of deleted.

Author: I believe @dav3r is correct. I ran into the same issue when I didn't include the `$(brew --prefix)/opt/gnu-getopt/bin`

Reviewer 2: > Agh you are right. It's keg-only on macOS and I have a file in my `.bashrc.d/` that adds it to the path with other keg-only formulae that I want to use. Fixing the PATH for `getopt` should be separate from `pyenv` environment configuration so this suggestion is still something I want, but it should be moved instead of completely deleted.

In full disclosure, @michaelsaki originally had the two `PATH` adjustments separate and I asked him to munge them.  _MY IGNORANCE ASTOUNDS AND AMAZES!!!_

Author: @jsf9k Obviously I was accounting for @mcdonnnj logic when I made the two `PATH`s 👀

Author: @mcdonnnj I addressed this here [495862a](https://github.com/cisagov/skeleton-generic/pull/166/commits/495862aeaf3430b0a69f92cf54b6415818139db2)

Otherwise, follow the steps below to manually configure your
environment.

#### Installing and using GNU `getopt`, `pyenv`, and `pyenv-virtualenv` ####

On the Mac, we recommend installing [brew](https://brew.sh/).  Then
installation is as simple as `brew install gnu-getopt pyenv pyenv-virtualenv` and
adding this to your profile:

```bash

> # Enable GNU getopt since it is keg-only

Reviewer 2: ```suggestion
# GNU getopt must be explicitly added to the path since it is
# keg-only
```

Otherwise, follow the steps below to manually configure your
environment.

#### Installing and using GNU `getopt`, `pyenv`, and `pyenv-virtualenv` ####

> On the Mac, we recommend installing [brew](https://brew.sh/).  Then

Reviewer: The more I look at it the more I hate "On the Mac". Please update all occurrences of "On the Mac" with "On macOS" and please update
https://github.com/cisagov/skeleton-generic/blob/324f6d426a149cb2c7bfc30ffcf7c1810b8ae1fa/setup-env#L100
to read "or macOS". There are at least a couple instances of "On the Mac" that I cannot provide suggestions for so apologies for that.

Author: See here: [0989d17](https://github.com/cisagov/skeleton-generic/pull/166/commits/0989d17f49a089e03e23b72ab638595d538add3f)

installation is as simple as `brew install gnu-getopt pyenv pyenv-virtualenv` and
adding this to your profile:

```bash
# GNU getopt must be explicitly added to the path since it is

> # keg-only

Reviewer 3: Help out folks who may not be super familiar with brew terminology:
```suggestion
# keg-only (https://docs.brew.sh/FAQ#what-does-keg-only-mean)
```

Otherwise, follow the steps below to manually configure your
environment.

#### Installing and using GNU `getopt`, `pyenv`, and `pyenv-virtualenv` ####

On the Mac, we recommend installing [brew](https://brew.sh/).  Then
installation is as simple as `brew install gnu-getopt pyenv pyenv-virtualenv` and
adding this to your profile:

> ```bash
> export PYENV_ROOT="$HOME/.pyenv"
> export PATH="$PYENV_ROOT/bin:$PATH"
> export PATH="$(brew --prefix)/opt/gnu-getopt/bin:$PATH"

Reviewer: I think I would prefer something like this to better separate and clarify what is going on.

```suggestion
```bash
# Enable GNU getopt since it is keg-only
export PATH="$(brew --prefix)/opt/gnu-getopt/bin:$PATH"

# Setup pyenv
export PYENV_ROOT="$HOME/.pyenv"
export PATH="$PYENV_ROOT/bin:$PATH"
```

Author: Makes sense to me

Otherwise, follow the steps below to manually configure your
environment.

#### Installing and using GNU `getopt`, `pyenv`, and `pyenv-virtualenv` ####

On macOS, we recommend installing [brew](https://brew.sh/).  Then
installation is as simple as `brew install gnu-getopt pyenv pyenv-virtualenv` and
adding this to your profile:

```bash
# GNU getopt must be explicitly added to the path since it is
# keg-only (https://docs.brew.sh/FAQ#what-does-keg-only-mean)
export PATH="$(brew --prefix)/opt/gnu-getopt/bin:$PATH"

# Setup pyenv
export PYENV_ROOT="$HOME/.pyenv"
export PATH="$PYENV_ROOT/bin:$PATH"
eval "$(pyenv init --path)"
eval "$(pyenv init -)"
eval "$(pyenv virtualenv-init -)"
```

> For Linux, Windows Subsystem for Linux (WSL), or on macOS (if you

Reviewer: ```suggestion
For Linux, Windows Subsystem for Linux (WSL), or macOS (if you
```

installed the prerequisites for your platform according to the
[`pyenv` wiki
page](https://github.com/pyenv/pyenv/wiki/common-build-problems).
`gnu-getopt` is generally included in the core utilities of most

> Linux distributions.

Reviewer: Again clearly separating the actual requirement from the Homebrew formula. Also make it clear what package GNU `getopt` would be included with in a Linux distribution.

```suggestion
GNU `getopt` is included in most Linux distributions as part of the
[`util-linux`](https://github.com/util-linux/util-linux) package.
```

installed the prerequisites for your platform according to the
[`pyenv` wiki
page](https://github.com/pyenv/pyenv/wiki/common-build-problems).

> `Gnu-getopt` is generally included in the core utilities of most

Reviewer 2: Fix capitalization:
```suggestion
`gnu-getopt` is generally included in the core utilities of most
```
